File: app/src/main/java/com/tuvakov/zeyoube/android/ui/feed/VideoFeedActivity.java
package com.tuvakov.zeyoube.android.ui.feed;

import android.Manifest;
import android.accounts.AccountManager;
import android.app.Dialog;
import android.content.Context;
import android.content.Intent;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.os.Bundle;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.gms.common.ConnectionResult;
import com.google.android.gms.common.GoogleApiAvailability;
import com.google.api.client.googleapis.extensions.android.gms.auth.GoogleAccountCredential;
import com.google.api.client.googleapis.extensions.android.gms.auth.GooglePlayServicesAvailabilityIOException;
import com.google.api.client.googleapis.extensions.android.gms.auth.UserRecoverableAuthIOException;
import com.tuvakov.zeyoube.android.R;
import com.tuvakov.zeyoube.android.VideoFeedSyncService;
import com.tuvakov.zeyoube.android.ZeYouBe;
import com.tuvakov.zeyoube.android.data.SyncStatus;
import com.tuvakov.zeyoube.android.ui.player.PlayerActivity;
import com.tuvakov.zeyoube.android.utils.DateTimeUtils;
import com.tuvakov.zeyoube.android.utils.PrefUtils;
import com.tuvakov.zeyoube.android.utils.YouTubeApiUtils;

import java.util.List;
import java.util.Objects;

import javax.inject.Inject;

import pub.devrel.easypermissions.AfterPermissionGranted;
import pub.devrel.easypermissions.EasyPermissions;

public class VideoFeedActivity extends AppCompatActivity
        implements EasyPermissions.PermissionCallbacks {

    private static final int REQUEST_ACCOUNT_PICKER = 1000;
    private static final int REQUEST_AUTHORIZATION = 1001;
    private static final int REQUEST_GOOGLE_PLAY_SERVICES = 1002;
    private static final int REQUEST_PERMISSION_GET_ACCOUNTS = 1003;

    private static final String TAG = "VideoFeedActivity";

    @Inject
    MainViewModelFactory mMainViewModelFactory;
    @Inject
    YouTubeApiUtils mYouTubeApiUtils;
    @Inject
    PrefUtils mPrefUtils;
    @Inject
    DateTimeUtils mDateTimeUtils;

    private GoogleAccountCredential mCredential;
    private MainViewModel mMainViewModel;

    /* Views */
    private RecyclerView mRecyclerView;
    private ProgressBar mProgressBar;
    private TextView mTextViewFeedback;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_video_feed);

        ((ZeYouBe) getApplication()).getAppComponent().injectVideoFeedActivityFields(this);

        mProgressBar = findViewById(R.id.progress_circular);
        mTextViewFeedback = findViewById(R.id.tv_feedback);

        /* Setup the RecyclerView */
        mRecyclerView = findViewById(R.id.rv_video_feed);
        mRecyclerView.setLayoutManager(new LinearLayoutManager(this));
        mRecyclerView.setHasFixedSize(true);

        VideoFeedAdapter videoFeedAdapter = new VideoFeedAdapter();
        videoFeedAdapter.setItemClickListener(video -> {
            Intent intent = new Intent(this, PlayerActivity.class);
            intent.putExtra(PlayerActivity.KEY_EXTRA_VIDEO_ID, video.getId());
            startActivity(intent);
        });

        mRecyclerView.setAdapter(videoFeedAdapter);

        /* Setup the ViewModel and start observing */
        mMainViewModel = new ViewModelProvider(this, mMainViewModelFactory)
                .get(MainViewModel.class);
        mMainViewModel.getVideoFeed().observe(this, videos -> {

            if (!hasContactsPermission()) {
                showMessage(R.string.msg_permission_get_accounts, View.GONE);
                return;
            }

            if (mMainViewModel.isSyncing()) {
                return;
            }

            if (videos.size() == 0) {
                showMessage(R.string.msg_info_empty_result_set, View.GONE);
                return;
            }

            videoFeedAdapter.submitList(videos);

            showRecyclerView();
            // Scroll to top after sync
            mRecyclerView.scrollToPosition(0);
        });

        /* Observe and react sync status */
        VideoFeedSyncService.STATUS.observe(this, this::handleSyncStatus);

        mCredential = mYouTubeApiUtils.getGoogleCredential();
        startSyncChain();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        switch (item.getItemId()) {
            case R.id.menu_item_delete_all_videos:
                clearUpData();
                return true;
            case R.id.menu_item_update_video_feed:
                if (mMainViewModel.isSyncing()) {
                    Toast.makeText(this, R.string.msg_info_already_syncing,
                            Toast.LENGTH_SHORT).show();
                    return true;
                } else if (!hasDayPassed()) {
                    Toast.makeText(this, getString(R.string.msg_info_sync_not_allowed),
                            Toast.LENGTH_LONG).show();
                    return true;
                }
                startSyncChain();
                return true;
            default:
                return true;
        }
    }

    /**
     * Called when an activity launched here (specifically, AccountPicker
     * and authorization) exits, giving you the requestCode you started it with,
     * the resultCode it returned, and any additional data from it.
     *
     * @param requestCode code indicating which activity result is incoming.
     * @param resultCode  code indicating the result of the incoming
     *                    activity result.
     * @param data        Intent (containing result data) returned by incoming
     *                    activity result.
     */
    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        switch (requestCode) {
            case REQUEST_GOOGLE_PLAY_SERVICES:
                if (resultCode != RESULT_OK) {
                    Toast.makeText(this, R.string.msg_warning_install_google_play_services,
                            Toast.LENGTH_SHORT).show();
                } else {
                    startSyncChain();
                }
                break;
            case REQUEST_ACCOUNT_PICKER:
                if (resultCode == RESULT_OK && data != null && data.getExtras() != null) {
                    String accountName = data.getStringExtra(AccountManager.KEY_ACCOUNT_NAME);
                    if (accountName != null) {
                        mPrefUtils.saveAccountName(accountName);
                        startSyncChain();
                    }
                }
                break;
            case REQUEST_AUTHORIZATION:
                if (resultCode == RESULT_OK) {
                    startSyncChain();
                } else {
                    showMessage(R.string.msg_permission_get_accounts, View.GONE);
                }
                break;
        }
    }

    /**
     * Respond to requests for permissions at runtime for API 23 and above.
     *
     * @param requestCode  The request code passed in
     *                     requestPermissions(android.app.Activity, String, int, String[])
     * @param permissions  The requested permissions. Never null.
     * @param grantResults The grant results for the corresponding permissions
     *                     which is either PERMISSION_GRANTED or PERMISSION_DENIED. Never null.
     */
    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions,
                                           @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        EasyPermissions.onRequestPermissionsResult(requestCode, permissions,
                grantResults, this);
    }

    @Override
    public void onPermissionsGranted(int requestCode, List<String> perms) {
        Log.d(TAG, "onPermissionsGranted: ");
//        showMessage(R.string.msg_info_restart_or_man_sync, View.GONE);
    }

    @Override
    public void onPermissionsDenied(int requestCode, List<String> perms) {
        Log.d(TAG, "onPermissionsDenied: ");
        showMessage(R.string.msg_permission_get_accounts, View.GONE);
    }

    /**
     * Attempt to call the API, after verifying that all the preconditions are
     * satisfied. The preconditions are: Google Play Services installed, an
     * account was selected and the device currently has online access. If any
     * of the preconditions are not satisfied, the app will prompt the user as
     * appropriate.
     */
    private void startSyncChain() {

        if (!isGooglePlayServicesAvailable()) {
            acquireGooglePlayServices();
        } else if (mPrefUtils.getAccountName() == null) {
            chooseAccount();
        } else if (!isDeviceOnline()) {
            showMessage(R.string.msg_warning_no_network, View.GONE);
        } else if (hasDayPassed()) {
            mMainViewModel.setIsSyncing(true);
            Intent intent = new Intent(this, VideoFeedSyncService.class);
            startService(intent);
        }
    }

    /**
     * Attempts to set the account used with the API credentials. If an account
     * name was previously saved it will use that one; otherwise an account
     * picker dialog will be shown to the user. Note that the setting the
     * account to use with the credentials object requires the app to have the
     * GET_ACCOUNTS permission, which is requested here if it is not already
     * present. The AfterPermissionGranted annotation indicates that this
     * function will be rerun automatically whenever the GET_ACCOUNTS permission
     * is granted.
     */
    @AfterPermissionGranted(REQUEST_PERMISSION_GET_ACCOUNTS)
    private void chooseAccount() {
        if (hasContactsPermission()) {
            String accountName = mPrefUtils.getAccountName();
            if (accountName != null) {
                startSyncChain();
            } else {
                // Start a dialog from which the user can choose an account
                startActivityForResult(
                        mCredential.newChooseAccountIntent(),
                        REQUEST_ACCOUNT_PICKER
                );
            }
        } else {
            // Request the GET_ACCOUNTS permission via a user dialog
            EasyPermissions.requestPermissions(
                    this,
                    getString(R.string.msg_permission_get_accounts),
                    REQUEST_PERMISSION_GET_ACCOUNTS,
                    Manifest.permission.GET_ACCOUNTS
            );
        }
    }

    private boolean isDeviceOnline() {
        ConnectivityManager connectivityManager =
                (ConnectivityManager) getSystemService(Context.CONNECTIVITY_SERVICE);
        NetworkInfo networkInfo = Objects.requireNonNull(connectivityManager).getActiveNetworkInfo();
        return (networkInfo != null && networkInfo.isConnected());
    }

    /**
     * Check that Google Play services APK is installed and up to date.
     *
     * @return true if Google Play Services is available and up to
     * date on this device; false otherwise.
     */
    private boolean isGooglePlayServicesAvailable() {
        GoogleApiAvailability apiAvailability =
                GoogleApiAvailability.getInstance();
        final int connectionStatusCode = apiAvailability.isGooglePlayServicesAvailable(this);
        return connectionStatusCode == ConnectionResult.SUCCESS;
    }

    /**
     * Attempt to resolve a missing, out-of-date, invalid or disabled Google
     * Play Services installation via a user dialog, if possible.
     */
    private void acquireGooglePlayServices() {
        GoogleApiAvailability apiAvailability = GoogleApiAvailability.getInstance();
        final int connectionStatusCode = apiAvailability.isGooglePlayServicesAvailable(this);
        if (apiAvailability.isUserResolvableError(connectionStatusCode)) {
            showGooglePlayServicesAvailabilityErrorDialog(connectionStatusCode);
        }
    }

    /**
     * Display an error dialog showing that Google Play Services is missing
     * or out of date.
     *
     * @param connectionStatusCode code describing the presence (or lack of)
     *                             Google Play Services on this device.
     */
    private void showGooglePlayServicesAvailabilityErrorDialog(final int connectionStatusCode) {
        GoogleApiAvailability apiAvailability = GoogleApiAvailability.getInstance();
        Dialog dialog = apiAvailability.getErrorDialog(
                VideoFeedActivity.this,
                connectionStatusCode,
                REQUEST_GOOGLE_PLAY_SERVICES
        );
        dialog.show();
    }

    private void handleSyncStatus(SyncStatus status) {
        int resultCode = status.getResultCode();
        if (resultCode == VideoFeedSyncService.STATUS_SYNC_STARTED) {
            showMessage(R.string.msg_info_sync_start, View.VISIBLE);
            return;
        }

        if (!hasContactsPermission()) {
            setStatusIdle();
            showMessage(R.string.msg_permission_get_accounts, View.GONE);
            clearUpData();
        }

        switch (resultCode) {
            case VideoFeedSyncService.STATUS_SYNC_SUCCESS:
                setStatusIdle();
                showRecyclerView();
                break;
            case VideoFeedSyncService.STATUS_SYNC_GOOGLE_PLAY_FAILURE:
                int code = ((GooglePlayServicesAvailabilityIOException) status.getException())
                        .getConnectionStatusCode();
                showGooglePlayServicesAvailabilityErrorDialog(code);
                setStatusIdle();
                break;
            case VideoFeedSyncService.STATUS_SYNC_AUTH_FAILURE:
                Intent intent = ((UserRecoverableAuthIOException) status.getException()).getIntent();
                startActivityForResult(intent, REQUEST_AUTHORIZATION);
                setStatusIdle();
                break;
            case VideoFeedSyncService.STATUS_SYNC_FAILURE:
                showMessage(R.string.msg_error_sync_failure, View.GONE);
                setStatusIdle();
                break;
        }

        mMainViewModel.setIsSyncing(false);
    }

    private void setStatusIdle() {
        VideoFeedSyncService.STATUS.setValue(
                new SyncStatus(VideoFeedSyncService.STATUS_SYNC_IDLE)
        );
    }

    private void showRecyclerView() {
        mProgressBar.setVisibility(View.GONE);
        mTextViewFeedback.setVisibility(View.GONE);
        mRecyclerView.setVisibility(View.VISIBLE);
    }

    private void showMessage(int messageStringId, int progressBarVisibility) {
        mRecyclerView.setVisibility(View.GONE);
        mProgressBar.setVisibility(progressBarVisibility);
        mTextViewFeedback.setVisibility(View.VISIBLE);
        mTextViewFeedback.setText(messageStringId);
    }

    private boolean hasContactsPermission() {
        return EasyPermissions.hasPermissions(this, Manifest.permission.GET_ACCOUNTS);
    }

    private boolean hasDayPassed() {
        return mDateTimeUtils.hasDayPassed(mPrefUtils.getLastSyncTime());
    }

    private void clearUpData() {
        mPrefUtils.deleteAccountName();
        mPrefUtils.saveLastSyncTime(0);
        mMainViewModel.deleteAllVideos();
        mMainViewModel.deleteAllSubscriptions();
    }
}

File: app/src/main/java/com/tuvakov/zetube/android/ui/feed/VideoFeedActivity.kt
package com.tuvakov.zetube.android.ui.feed

import android.Manifest
import android.accounts.AccountManager
import android.app.Activity
import android.content.Context
import android.content.DialogInterface
import android.content.Intent
import android.net.ConnectivityManager
import android.os.Bundle
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.ActionBarDrawerToggle
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.GravityCompat
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.RecyclerView
import com.google.android.gms.common.ConnectionResult
import com.google.android.gms.common.GoogleApiAvailability
import com.google.android.material.navigation.NavigationView
import com.google.api.client.googleapis.extensions.android.gms.auth.GoogleAccountCredential
import com.google.api.client.googleapis.extensions.android.gms.auth.GooglePlayServicesAvailabilityIOException
import com.google.api.client.googleapis.extensions.android.gms.auth.UserRecoverableAuthIOException
import com.tuvakov.zetube.android.R
import com.tuvakov.zetube.android.ZeTubeApp
import com.tuvakov.zetube.android.data.Video
import com.tuvakov.zetube.android.databinding.ActivityVideoFeedBinding
import com.tuvakov.zetube.android.ui.channeldetail.*
import com.tuvakov.zetube.android.ui.channels.ChannelsActivity
import com.tuvakov.zetube.android.utils.*
import pub.devrel.easypermissions.AfterPermissionGranted
import pub.devrel.easypermissions.EasyPermissions
import pub.devrel.easypermissions.EasyPermissions.PermissionCallbacks
import java.util.*
import javax.inject.Inject

class VideoFeedActivity : AppCompatActivity(),
        PermissionCallbacks,
        NavigationView.OnNavigationItemSelectedListener {

    @Inject
    lateinit var mViewModelFactory: ViewModelFactory

    @Inject
    lateinit var mYouTubeApiUtils: YouTubeApiUtils

    @Inject
    lateinit var mPrefUtils: PrefUtils

    @Inject
    lateinit var mDateTimeUtils: DateTimeUtils

    private lateinit var mCredential: GoogleAccountCredential
    private lateinit var mMainViewModel: MainViewModel
    private lateinit var binding: ActivityVideoFeedBinding
    private lateinit var mTextViewAccountName: TextView

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityVideoFeedBinding.inflate(layoutInflater)
        setContentView(binding.root)

        // Inject fields
        (application as ZeTubeApp).appComponent.injectVideoFeedActivityFields(this)

        setSupportActionBar(binding.lToolbar.toolbar)

        /* Setup navigation and drawer */
        binding.navView.setNavigationItemSelectedListener(this)
        mTextViewAccountName = binding.navView.getHeaderView(0).findViewById(R.id.tv_nav_account)
        setupDrawer()

        /* Setup the RecyclerView */
        val videoFeedAdapter = VideoFeedAdapter().also {
            it.registerAdapterDataObserver(object : RecyclerView.AdapterDataObserver() {
                override fun onChanged() {
                    binding.lVideoFeed.rvVideoFeed.scrollToPosition(0)
                }

                override fun onItemRangeRemoved(positionStart: Int, itemCount: Int) {
                    binding.lVideoFeed.rvVideoFeed.scrollToPosition(0)
                }

                override fun onItemRangeMoved(fromPosition: Int, toPosition: Int, itemCount: Int) {
                    binding.lVideoFeed.rvVideoFeed.scrollToPosition(0)
                }

                override fun onItemRangeInserted(positionStart: Int, itemCount: Int) {
                    binding.lVideoFeed.rvVideoFeed.scrollToPosition(0)
                }

                override fun onItemRangeChanged(positionStart: Int, itemCount: Int) {
                    binding.lVideoFeed.rvVideoFeed.scrollToPosition(0)
                }

                override fun onItemRangeChanged(positionStart: Int, itemCount: Int, payload: Any?) {
                    binding.lVideoFeed.rvVideoFeed.scrollToPosition(0)
                }
            })
        }
        binding.lVideoFeed.rvVideoFeed.adapter = videoFeedAdapter

        /* Setup the ViewModel and start observing */
        mMainViewModel = ViewModelProvider(this, mViewModelFactory)
                .get(MainViewModel::class.java)

        mMainViewModel.videoFeed.observe(this, Observer { videos: List<Video> ->
            if (videos.isEmpty() && mMainViewModel.isSuccess) {
                mMainViewModel.setEmptyListStatus()
            }
            videoFeedAdapter.submitList(videos)
        })

        /* Observe and react sync status */
        mMainViewModel.status.observe(this, Observer { status: LiveDataState ->
            handleSyncStatus(status)
        })

        mCredential = mYouTubeApiUtils.googleCredential

        startSyncChain()
    }

    override fun onBackPressed() {
        if (binding.layoutDrawer.isDrawerOpen(GravityCompat.START)) {
            binding.layoutDrawer.closeDrawer(GravityCompat.START)
        } else {
            super.onBackPressed()
        }
    }

    override fun onNavigationItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.nav_feed -> mMainViewModel.loadAllVideos()
            R.id.nav_channels -> gotoChannels()
            R.id.nav_saved_videos -> mMainViewModel.loadSavedVideos()
        }
        binding.layoutDrawer.closeDrawer(GravityCompat.START)
        return true
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_main, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.menu_item_delete_all_videos -> {
                showDeleteDialog()
                true
            }
            R.id.menu_item_update_video_feed -> {
                if (mMainViewModel.isSyncing) {
                    Toast.makeText(this, R.string.msg_info_already_syncing,
                            Toast.LENGTH_SHORT).show()
                    return true
                } else if (!isSyncAllowed()) {
                    Toast.makeText(
                            this,
                            getString(R.string.msg_info_sync_not_allowed, SyncUtils.SYNC_INTERVAL_HOURS),
                            Toast.LENGTH_LONG).show()
                    return true
                }
                startSyncChain()
                true
            }
            else -> true
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        when (requestCode) {
            REQUEST_GOOGLE_PLAY_SERVICES -> {
                if (resultCode != Activity.RESULT_OK) {
                    Toast.makeText(this, R.string.msg_warning_install_google_play_services,
                            Toast.LENGTH_SHORT).show()
                } else {
                    startSyncChain()
                }
            }
            REQUEST_ACCOUNT_PICKER -> {
                if (resultCode == Activity.RESULT_OK && data != null && data.extras != null) {
                    val accountName = data.getStringExtra(AccountManager.KEY_ACCOUNT_NAME)
                    if (accountName != null) {
                        mPrefUtils.saveAccountName(accountName)
                        startSyncChain()
                    }
                }
            }
            REQUEST_AUTHORIZATION -> {
                if (resultCode == Activity.RESULT_OK) {
                    startSyncChain()
                } else {
                    showMessage(R.string.msg_permission_get_accounts, View.GONE)
                }
            }
        }
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>,
                                            grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        EasyPermissions.onRequestPermissionsResult(requestCode, permissions,
                grantResults, this)
    }

    override fun onPermissionsGranted(requestCode: Int, perms: List<String>) {
        Log.d(TAG, "onPermissionsGranted: ")
//        showMessage(R.string.msg_info_restart_or_man_sync, View.GONE);
    }

    override fun onPermissionsDenied(requestCode: Int, perms: List<String>) {
        Log.d(TAG, "onPermissionsDenied: ")
        showMessage(R.string.msg_permission_get_accounts, View.GONE)
    }

    private fun startSyncChain() {
        if (!hasContactsPermission()) {
            showMessage(R.string.msg_permission_get_accounts, View.GONE)
            clearUpData()
        }
        if (!isGooglePlayServicesAvailable) {
            acquireGooglePlayServices()
        } else if (mPrefUtils.accountName == null) {
            chooseAccount()
        } else if (!isDeviceOnline) {
            showMessage(R.string.msg_warning_no_network, View.GONE)
        } else if (isSyncAllowed()) {
            mMainViewModel.sync()
        }
    }

    @AfterPermissionGranted(REQUEST_PERMISSION_GET_ACCOUNTS)
    private fun chooseAccount() {
        if (hasContactsPermission()) {
            val accountName = mPrefUtils.accountName
            if (accountName != null) {
                startSyncChain()
            } else {
                // Start a dialog from which the user can choose an account
                startActivityForResult(
                        mCredential.newChooseAccountIntent(),
                        REQUEST_ACCOUNT_PICKER
                )
            }
        } else {
            // Request the GET_ACCOUNTS permission via a user dialog
            EasyPermissions.requestPermissions(
                    this,
                    getString(R.string.msg_permission_get_accounts),
                    REQUEST_PERMISSION_GET_ACCOUNTS,
                    Manifest.permission.GET_ACCOUNTS
            )
        }
    }

    private val isDeviceOnline: Boolean
        get() {
            val connectivityManager = getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
            val networkInfo = Objects.requireNonNull(connectivityManager).activeNetworkInfo
            return networkInfo != null && networkInfo.isConnected
        }

    private val isGooglePlayServicesAvailable: Boolean
        get() {
            val apiAvailability = GoogleApiAvailability.getInstance()
            val connectionStatusCode = apiAvailability.isGooglePlayServicesAvailable(this)
            return connectionStatusCode == ConnectionResult.SUCCESS
        }

    private fun acquireGooglePlayServices() {
        val apiAvailability = GoogleApiAvailability.getInstance()
        val connectionStatusCode = apiAvailability.isGooglePlayServicesAvailable(this)
        if (apiAvailability.isUserResolvableError(connectionStatusCode)) {
            showGooglePlayServicesAvailabilityErrorDialog(connectionStatusCode)
        }
    }

    private fun showGooglePlayServicesAvailabilityErrorDialog(connectionStatusCode: Int) {
        val apiAvailability = GoogleApiAvailability.getInstance()
        val dialog = apiAvailability.getErrorDialog(
                this@VideoFeedActivity,
                connectionStatusCode,
                REQUEST_GOOGLE_PLAY_SERVICES
        )
        dialog.show()
    }

    private fun handleSyncStatus(status: LiveDataState) {
        Log.d(TAG, status.toString())
        when (status) {
            InProgress -> showMessage(R.string.msg_info_sync_start, View.VISIBLE)
            Success -> showRecyclerView()
            EmptyList -> showMessage(R.string.msg_info_empty_result_set, View.GONE)
            is Error -> {
                Log.e(TAG, status.exception?.message, status.exception!!)
                when (status.exception) {
                    is ImmatureSyncException -> {
                        TODO()
                    }
                    is GooglePlayServicesAvailabilityIOException -> {
                        val code = status.exception.connectionStatusCode
                        showGooglePlayServicesAvailabilityErrorDialog(code)
                    }
                    is UserRecoverableAuthIOException -> {
                        val intent = status.exception.intent
                        startActivityForResult(intent, REQUEST_AUTHORIZATION)
                    }
                    else -> {
                        showMessage(R.string.msg_error_sync_failure, View.GONE)
                    }
                }
            }
        }
    }

    private fun setupDrawer() {
        val toggle = ActionBarDrawerToggle(this, binding.layoutDrawer,
                binding.lToolbar.toolbar, R.string.navigation_drawer_open,
                R.string.navigation_drawer_close
        )
        binding.layoutDrawer.addDrawerListener(toggle)
        toggle.syncState()

        /* Set up account name in navigation header */
        val accountName = mPrefUtils.accountName
        mTextViewAccountName.text = accountName
    }

    private fun gotoChannels() {
        if (mPrefUtils.accountName != null && !mMainViewModel.isSyncing) {
            val intent = Intent(this, ChannelsActivity::class.java)
            startActivity(intent)
        }
    }

    private fun showRecyclerView() {
        with(binding.lVideoFeed) {
            progressCircular.hide()
            tvFeedback.hide()
            rvVideoFeed.show()
        }
    }

    private fun showMessage(messageStringId: Int, progressBarVisibility: Int = View.GONE) {
        with(binding.lVideoFeed) {
            rvVideoFeed.hide()
            tvFeedback.show()
            tvFeedback.setText(messageStringId)
            progressCircular.visibility = progressBarVisibility

        }
    }

    private fun hasContactsPermission(): Boolean {
        return EasyPermissions.hasPermissions(this, Manifest.permission.GET_ACCOUNTS)
    }

    private fun isSyncAllowed(): Boolean {
        return mDateTimeUtils.isSyncAllowed(mPrefUtils.lastSyncTime)
    }

    private fun clearUpData() {
        mPrefUtils.emptyPreferences()
        mMainViewModel.emptyDatabase()
        /* Set account name text view empty */
        mTextViewAccountName.text = ""
    }

    private fun showDeleteDialog() {
        /* Don't show the dialog if there is no data */
        if (mPrefUtils.accountName == null && mPrefUtils.lastSyncTime == 0L) {
            Toast.makeText(this, R.string.msg_no_data_to_delete, Toast.LENGTH_SHORT).show()
            return
        }
        val builder = AlertDialog.Builder(this, R.style.ZeTubeAlertDialog)
        builder.setMessage(R.string.msg_delete_dialog)
        builder.setCancelable(true)
        builder.setPositiveButton(R.string.btn_txt_confirm) { _: DialogInterface?, _: Int -> clearUpData() }
        builder.setNegativeButton(R.string.btn_txt_cancel) { dialog: DialogInterface, _: Int -> dialog.cancel() }
        builder.create().show()
    }

    companion object {
        private const val REQUEST_ACCOUNT_PICKER = 1000
        private const val REQUEST_AUTHORIZATION = 1001
        private const val REQUEST_GOOGLE_PLAY_SERVICES = 1002
        private const val REQUEST_PERMISSION_GET_ACCOUNTS = 1003
        private const val TAG = "VideoFeedActivity"
    }
}